prov/gni/src/gnix_vc.c

```diff
@@ -86,6 +86,94 @@ static inline void __gnix_vc_set_ht_key(void *gnix_addr,
 	*key = *((gnix_ht_key_t *)gnix_addr);
 }
 
+struct gnix_vc *_gnix_ep_vc_lookup(struct gnix_fid_ep *ep, uint64_t key)
+{
+	struct gnix_vc *vc;
+	int ret;
+
+	assert(ep->av);
+
+	if (ep->av->type == FI_AV_TABLE) {
+		ret = _gnix_vec_at(ep->vc_table, (void **)&vc, key);
+		if (ret != FI_SUCCESS) {
+			vc = NULL;
+		}
+	} else {
+		vc = (struct gnix_vc *)_gnix_ht_lookup(ep->vc_ht, key);
+	}
+
+	return vc;
+}
+
+int _gnix_ep_vc_store(struct gnix_fid_ep *ep, struct gnix_vc *vc, uint64_t key)
+{
+	int ret;
+
+	assert(ep->av);
+
+	if (ep->av->type == FI_AV_TABLE) {
+		ret = _gnix_vec_insert_at(ep->vc_table, (void *)vc, key);
+	} else {
+		ret = _gnix_ht_insert(ep->vc_ht, key, vc);
+	}
+
+	return ret;
+}
+
+static int __gnix_vc_gnix_addr_equal(struct dlist_entry *item, const void *arg)
+{
+	struct gnix_vc *vc = dlist_entry(item, struct gnix_vc, list);
+
+	return GNIX_ADDR_EQUAL(vc->peer_addr, *(struct gnix_address *)arg);
+}
+
+/* Find an unmapped VC that matches 'dest_addr' and map it into the EP's VC
+ * look up table.  VC lock must be held. */
+static struct gnix_vc *__gnix_vc_lookup_unmapped(struct gnix_fid_ep *ep,
+						 fi_addr_t dest_addr)
+{
+	struct gnix_av_addr_entry *av_entry;
+	struct dlist_entry *entry;
+	struct gnix_vc *vc;
+	int ret;
+
+	/* Determine if the fi_addr now exists in the AV. */
+	ret = _gnix_av_lookup(ep->av, dest_addr, &av_entry);
+	if (ret != FI_SUCCESS) {
+		GNIX_WARN(FI_LOG_EP_DATA,
+			  "_gnix_av_lookup for addr 0x%lx returned %s\n",
+			  dest_addr, fi_strerror(-ret));
+		return NULL;
+	}
+
+	/* Find a pre-existing, unmapped VC that matches the gnix_address
+	 * mapped by dest_addr. */
+	entry = dlist_remove_first_match(&ep->unmapped_vcs,
+					 __gnix_vc_gnix_addr_equal,
+					 (void *)&av_entry->gnix_addr);
+	if (entry) {
+		/* Found a matching, unmapped VC.  Map dest_addr to the VC in
+		 * the EP's VC look up table. */
+		vc = dlist_entry(entry, struct gnix_vc, list);
+		GNIX_INFO(FI_LOG_EP_CTRL,
+			  "Found unmapped VC: %p gnix_addr: 0x%lx fi_addr: 0x%lx\n",
+			  vc, vc->peer_addr, vc->peer_fi_addr);
+
+		ret = _gnix_ep_vc_store(ep, vc, dest_addr);
```

Comment:
I had seen this fail when using vector storage and hitting the upper limit.  The vector enforces a hard limit of entries.  The hashtable doesn't have a hard limit.  We might need to remove the vector size limit.  I added an arbitrarily large limit for now.  This shouldn't fail under normal conditions.